MetaMask/controllers
# Improve support for backfilling types via `.d.ts`'s

Currently, we exclude test files from TypeScript's purview, so that only
files in `src/` get emitted to `dist/`. This makes sense from a release
perspective, but it interferes with development.

One example of this interference involves [ambient modules][1]. We have
a file `dependencies.d.ts` which we can use to backfill missing types
for various dependencies. The problem is that while TypeScript (or, more
accurately, `tsserver`) can see the types specified here when reading
source files, it doesn't seem to pick them up for test files. This
causes in-editor inconsistencies with type analysis and, in some cases,
warnings such as:

* `import` can only be used with `esInteropModuleFlag: true`
* `X` is missing type definitions

The reason this is happening is that type definition files are not
intended to just be placed in the source directory. You can either use
triple-slash directives to bring them in, or you can modify a setting in
`tsconfig.json` (`paths`) that will [instruct TypeScript where it should
find types for modules that are imported][1]. This commit chooses the
latter option. Note that this requires we split up `dependencies.d.ts`
into separate type definition files, one per module.

In addition to modifying `tsconfig.json`, this commit also breaks up
this config file into a global/development version and a version that is
specific to the build/release process. This allows editors to use
the same exact TypeScript settings for test files as for non-test files.
`tsconfig.build.json` is used by `yarn build` and will limit the files
emitted to `dist/` to only non-test files in `src/`, as in the original
configuration.

[1]: https://www.typescriptlang.org/docs/handbook/modules.html#ambient-modules
[2]: https://www.typescriptlang.org/docs/handbook/module-resolution.html#path-mapping

---

## Manual testing steps

* Check out the `main` branch.
* Open `src/assets/AssetsContractController.test.ts`.
* Hover over `MAINNET_PROVIDER` on line 6. Note how the type is `any`. This is because the types for `ethjs-provider-http` is being stubbed in `src/dependencies.d.ts`.
* Update the `declare module 'ethjs-provider-http';` line in `src/dependencies.d.ts` to read:
    ```
    declare module 'ethjs-provider-http' {
      import type { JsonRpcRequest } from 'json-rpc-engine';
      export type EthQueryMethodCallback<R> = (error: any, response: R) => void;

      export default class HttpProvider {
        host: string;
        timeout: number;

        constructor(host: string, timeout?: number);

        sendAsync<P, R>(
          request: JsonRpcRequest<P>,
          callback: EthQueryMethodCallback<R>,
        ): void;
      }
    }
    ```
* Hover over `MAINNET_PROVIDER` in `src/assets/AssetsContractController.test.ts` again. Note how the type is still `any`.
* Remove your changes to `src/dependencies.d.ts`.
* Now check out this branch.
* Restore your changes to `src/dependencies.d.ts`.
* Hover over `MAINNET_PROVIDER` in `src/assets/AssetsContractController.test.ts` a third time. Note how the type is properly `HttpProvider`.
* Finally, run `yarn build`. All of the files we would expect to show up in `dist/` should be there (i.e. no extra `.d.ts` files, no tests, etc.)
